### ISSUE TITLE
Clarify that MVC components provided through WebMvcRegistrations are subject to subsequent processing and configuration by MVC

### ISSUE THREAD

**Issue by vpavic** [spring-projects/spring-boot#31232]
Previously, I was able to disable Spring MVC's trailing slash matching using something like this:

```java
@Configuration(proxyBeanMethods = false)
class WebMvcConfiguration {

    @Bean
    WebMvcRegistrations webMvcRegistrations() {
        return new WebMvcRegistrations() {

            @Override
            public RequestMappingHandlerMapping getRequestMappingHandlerMapping() {
                RequestMappingHandlerMapping handlerMapping = new RequestMappingHandlerMapping();
                handlerMapping.setUseTrailingSlashMatch(false);
                return handlerMapping;
            }

         };
    }

}
```

This doesn't work any more since `2.6.0` and the change of behavior appears to be related to introduction of `PathPatternParser` as the default path matching strategy.

I've put together a reproduced in https://github.com/vpavic/mcve-spring-mvc-trailing-slash.

By default, trailing slash matching is enabled and `curl -sv -u user:password http://localhost:8080/resources/` (note the trailing slash) returns `200` response. There are two profiles that disable trailing slash matching - `configurer` and `registrations`. The former uses `WebMvcConfigurer` based approach and this works (meaning, cURL request returns `404`), while the latter uses `WebMvcRegistrations` based approach (as in the snippet above) and this doesn't work (returns `200`).

Downgrading Spring Boot to `2.5.x` makes the `registrations` profile work as expected.

I'm reporting this here for starters because what's failing is a Spring Boot specific configuration mechanism (`WebMvcRegistrations`) even though it seems to me that the fix might very well be needed in the Framework.

**Comment by wilkinsona** [comment]
Thanks for the report and, in particular, the MCVE. 

`WebMvcRegistrations` plugs into Spring MVC via `WebMvcConfigurationSupport.createRequestMappingHandlerMapping()`. Its javadoc states that it is a "protected method for plugging in a custom subclass of `RequestMappingHandlerMapping`. Based on this, I would argue that your `WebMvcRegistrations` instance isn't using the plug point as intended. The javadoc of `WebMvcRegistrations.getRequestMappingHandlerMapping` also alludes to the current behaviour as it says that the method should "return the custom `RequestMappingHandlerMapping` that should be used and processed by the MVC configuration". The key part here is "and processed by the MVC configuration" as it's this subsequent processing that is causing the trailing slash configuration to be lost.

In short, your `WebMvcConfigurer`-based approach is using the extension points and configuration mechanisms as intended. We can use this issue to clarify the javadoc and reference documentation around `WebMvcRegistrations` to make it clear that it's intended for providing custom subclasses of MVC components and that configuration should be done via `WebMvcConfigurer`.

**Comment by vpavic** [comment]
OK, thanks. I suspected if a change was to be made it would be in Framework but it's good to also have additional clarifications in Boot's docs.

**Comment by vpavic** [comment]
I like the idea @quaff - I've put together a quick PR (see #31327) to consider it.

**Comment by rstoyanchev** [comment]
I agree that the current situation is a bit strange. The `useTrailingSlashMatch` option on RMHM tries to work as a convenience method by also setting the same on `PathPatternParser`, but the fact that `PathPatternParser` isn't set by default and must be set explicitly, in effect means that they work only if invoked in the correct order. 

For 6.0, I am proposing to use `PathPatternParser` by default in Spring Framework with https://github.com/spring-projects/spring-framework/issues/28607, which would remove the need for a `PathPatternParser` to be set explicitly. That would address the issue. I am also proposing to deprecate the trailingSlash option altogether in https://github.com/spring-projects/spring-framework/issues/28552#issuecomment-1152391820, but both of these pend team decisions.
